src/Serilog.Sinks.Loggly/Sinks/Loggly/FileBufferDataProvider.cs

```diff
@@ -0,0 +1,279 @@
+
+    /// <summary>
+    /// Provides a facade to all File operations, namely bookmark management and 
+    /// buffered data readings
+    /// </summary>
+    class FileBufferDataProvider : IBufferDataProvider
+    {
+        readonly string _candidateSearchPath;
+        readonly string _logFolder;
+        readonly string _bookmarkFilename;
+
+        readonly int _batchPostingLimit;
+        readonly long? _eventBodyLimitBytes;
+
+        readonly IFileSystemAdapter _fileSystemAdapter;
+        readonly IBookmarkProvider _bookmarkProvider;
+        readonly Encoding _encoding;
+
+        readonly JsonSerializer _serializer = JsonSerializer.Create();
+
+        // the following fields control the internal state and position of the queue
+        Bookmark _currentBookmark;
+        Bookmark _futureBookmark;
+        IEnumerable<LogglyEvent> _currentBatchOfEventsToProcess;
+
+        public FileBufferDataProvider(
+            string baseBufferFileName, 
+            IFileSystemAdapter fileSystemAdapter, 
+            IBookmarkProvider bookmarkProvider,
+            Encoding encoding,
+            int batchPostingLimit,
+            long? eventBodyLimitBytes
+        )
+        {
+            _bookmarkFilename = Path.GetFullPath(baseBufferFileName + ".bookmark");
+            _logFolder = Path.GetDirectoryName(_bookmarkFilename);
+            _candidateSearchPath = Path.GetFileName(baseBufferFileName) + "*.json";
+
+            _fileSystemAdapter = fileSystemAdapter;
+            _bookmarkProvider = bookmarkProvider;
+            _encoding = encoding;
+            _batchPostingLimit = batchPostingLimit;
+            _eventBodyLimitBytes = eventBodyLimitBytes;
+        }
+
+        public IEnumerable<LogglyEvent> GetBatchOfEvents()
+        {
+            //if current batch has not yet been processed, return it
+            if (_currentBatchOfEventsToProcess != null)
+                return _currentBatchOfEventsToProcess;
+
+            //if we have a bookmark in place, it may be the next position to read from
+            // otherwise try to get a valid one
+            if (_currentBookmark == null)
+            {
+                //read the current bookmark from file, and if invalid, try to create a valid one
+                _currentBookmark = TryGetValidBookmark();
+
+                if (!IsValidBookmark(_currentBookmark))
+                    return new List<LogglyEvent>();
+            }
+
+            //bookmark is valid, so lets get the next batch from the files.
+            _currentBatchOfEventsToProcess = GetListOfEvents(_currentBookmark.FileName);
+            //_futureBookmark = new Bookmark(nextByteCount, _currentBookmark.FileName); //performed in futureBookmark
+            return _currentBatchOfEventsToProcess;
+        }
+
+        public void MarkCurrentBatchAsProcessed()
+        {
+            //reset internal state: 
+            if(_futureBookmark != null)
+                _bookmarkProvider.UpdateBookmark(_futureBookmark);
```

Comment:
if the bookmark is null, shouldn't we need to update the file to reflect that as well?

Reply:
In this case, if we update the bookmark file (as in, clearing it out), The read op in `FileBasedBookmarkProvider.GetCurrentBookmarkPosition()`will return `null`which I believe is consistent with our expectations -> if null, the search will continue to whatever files are on disk and rebuild the bookmark if possible. The update op will need to support receiving a null `Bookmark`. I think that's ok. Do you agree?

Reply:
I was starting to look into this and remembered the issue I was having (#25) - I sometimes lost the bookmark position due to bad reads from the bookmark.  

In this case, It might be better to preserve the position (even if outdated) then to reset it, and avoid re-reading from the same buffer file (as was occurring in the issue). An outdated position / bookmark will not read any items, and based on intent, move on to the next file if it exists. Clearing out the contents may risk reintroducing the bad behaviour.